### PR TITLE
Implement ipni-sync http over libp2p

### PR DESCRIPTION
**Pull request by gammazero**
## Serve ipnisync HTTP over libp2p

This PR makes use of the new libp2phttp functionality for serving and requesting ipnisync over libp2p.

The new publisher can publish ipnisync-http over libp2p if supplied with a stream host, over plain HTTP if supplied with a listen address, or both. The publisher also works as an HTTP [handler](https://pkg.go.dev/net/http@go1.21.0#Handler) to allow it to be used with existing HTTP listeners.

The sync client works with ipnisync-http over libp2p, plain HTTP, and legacy data-transfer sync. It is able to detect which protocol and transport to use with any provider, new or old.

### Protocol negotiation
When a `Syncer` is created to sync with a specific advertisement publisher, it first uses libp2phttp to negotiate with the publisher whether to use HTTP with or without libp2p. If the publisher is an older publisher that does not support this negotiation, then the client tries to use plain HTTP or the legacy data-transfer/graphsync protocol depending on the publisher address.

If a publisher is newer, but uses an existing HTTP server, it can still support protocol negotiation by supporting the `/.well-known/` HTTP endpoint. [Here](https://gist.github.com/MarcoPolo/3fb6b479b25d8a8e43eaadc516c60c29) is an example of adding support for this. This is optional, and the cost for not supporting it is one additional HTTP round trip between the indexer and publisher per sync operation.

### API Changes
This PR has some minor `ipnisync` API changes to be aware of.

#### Publisher API Changes:
- The `NewPublisher` function does not take a single HTTP listen address as its first argument. It now takes a list of HTTP address as an option, `WithHTTPListenAddrs`, since HTTP listen addresses are not required when using a libp2p stream host or with an existing HTTP server.
- The `NewPublisher` can be given a libp2p stream  host to serve advertisements, by specifying the option `WithStreamHost`.
- The `dagsync/p2p/protocol/head` package has moved to `dagsync/dtsync/head` since it is only used in `dtsync`. This should not affect any external applications.
- The `WithServer` has been renamed to `WithStartServer`.
-  A new option, `WithRequireTLS` tells whether to require https or allow the publisher to serve non-secure http. Default is false, allowing non-secure HTTP.

#### Sync client API changes:
- The `NewSync` function does not take an `http.Client`. Options are used to configure a default or retryable HTTP client.
- New options are available to the `NewSync` function:
  - `ClientAuthServerPeerID` tells the sync client that it must authenticate the Server's PeerID.
  - `ClientHTTPTimeout` specifies a time limit for HTTP requests
  - `ClientStreamHost` specifies an optional stream based libp2p host
  - `ClientHTTPRetry` configures a retriable HTTP client
- The `NewSyncer` function now takes a single `peer.AddrInfo` instead of a separate `peer.ID` and `[]multiaddr.Multiaddr` arguments.

#### `dagsync.Subscriber` API changes:
- `NewSubscriber` no longer has a `HttpClient` option. It now supports the following options:
  - `HttpTimeout` which is passed through to the the `ClientHttpTimeout` option
  - `RetryableHTTPClient` which is passed through to the `ClientHTTPRetry` option.

### CODE REVIEW

**Comment by gammazero** [comment]
From @MarcoPolo in #103, [here](https://github.com/ipni/go-libipni/pull/103#issuecomment-1675231219)
I took a look at Frisbii, I think there's only one thing to add if you want to support the .well-known/libp2p endpoint. Which you probably should unless you expect every client to be able to learn about your protocols and their prefixes out-of-band (which is fine! just more work). The change is small since you already have go-libp2p as a dependency, only 4 lines of code. Here's the patch: https://gist.github.com/MarcoPolo/3fb6b479b25d8a8e43eaadc516c60c29.

File: dagsync/ipnisync/sync.go [hunk 1]
```diff
@@ -20,67 +23,155 @@ import (
...
+	"github.com/ipni/go-libipni/mautil"
 	"github.com/libp2p/go-libp2p/core/peer"
-	"github.com/multiformats/go-multiaddr"
+	"github.com/libp2p/go-libp2p/core/protocol"
+	libp2phttp "github.com/libp2p/go-libp2p/p2p/http"
 	"github.com/multiformats/go-multihash"
 )
 
-const defaultHttpTimeout = 10 * time.Second
+const ProtocolID = protocol.ID("/ipnisync/v1")
```
Comment (masih) [inline dagsync/ipnisync/sync.go:33]: Too late to change this to `/ipni/sync/v1`?

**Review by masih — APPROVED** [review]
🚀

File: dagsync/ipnisync/sync.go (hunk 1 above)
Comment (gammazero) [reply dagsync/ipnisync/sync.go:33]: Would it be better to match the HTTP path, which is `/ipni/v1/ad/` ?

Comment (masih) [reply dagsync/ipnisync/sync.go:33]: Not sure http path as libp2p protocol ID makes sense? That protocol exposes multiple paths right?

Comment (gammazero) [reply dagsync/ipnisync/sync.go:33]: Only one path. As per spec the only path is `/ipni/v1/ad/`, which is followed by the resource `head` or the CID to fetch.

Comment (gammazero) [reply dagsync/ipnisync/sync.go]: I decided to use the IPNIPath as the protocol ID. I provided an explanation in the [comment here](https://github.com/ipni/go-libipni/pull/113/files#diff-ea78075c5ae0828f3c63371f8668ed407bd104216f299f45b8a7e4cfd630282eR10-R14).

**Pull request merged**